File: Scheduling/Weighted interval scheduling/WIS.py
def weighted_interval_scheduling(intervals):
    # Sort intervals by their end time
    intervals.sort(key=lambda x: x[1])
    
    # Create an array to store the maximum weight of each interval
    # where max_weight[i] represents the maximum weight of the interval
    # set that ends at i
    max_weight = [0] * len(intervals)
    
    # Initialize the first maximum weight to be the weight of the first interval
    max_weight[0] = intervals[0][2]
    
    # Calculate the maximum weight for each interval using dynamic programming
    for i in range(1, len(intervals)):
        # Find the latest non-overlapping interval j
        j = i - 1
        while j >= 0 and intervals[j][1] > intervals[i][0]:
            j -= 1
        
        # Calculate the maximum weight of the current interval
        max_weight[i] = max(intervals[i][2] + max_weight[j], max_weight[i-1])
    
    # Return the maximum weight of the last interval
    return max_weight[-1]


# Get the input from the user
n = int(input("Enter the number of intervals: "))
intervals = []
for i in range(n):
    start, end, weight = map(int, input(f"Enter the start time, end time, and weight of interval {i+1}: ").split())
    intervals.append((start, end, weight))

# Call the function to get the maximum total weight of non-overlapping intervals
max_weight = weighted_interval_scheduling(intervals)

# Print the result
print("The maximum total weight of non-overlapping intervals is:", max_weight)

File: Scheduling/Weighted interval scheduling/WIS.java
import java.util.*;

public class WeightedIntervalScheduling {

    static class Interval implements Comparable<Interval> {
        int start;
        int end;
        int weight;
        
        public Interval(int start, int end, int weight) {
            this.start = start;
            this.end = end;
            this.weight = weight;
        }
        
        @Override
        public int compareTo(Interval other) {
            return this.end - other.end;
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<Interval> intervals = new ArrayList<>();
        System.out.print("Enter the number of intervals: ");
        int n = scanner.nextInt();
        for (int i = 0; i < n; i++) {
            System.out.print("Enter the start time of interval " + (i+1) + ": ");
            int start = scanner.nextInt();
            System.out.print("Enter the end time of interval " + (i+1) + ": ");
            int end = scanner.nextInt();
            System.out.print("Enter the weight of interval " + (i+1) + ": ");
            int weight = scanner.nextInt();
            intervals.add(new Interval(start, end, weight));
        }
        int maxWeight = findMaxWeight(intervals);
        System.out.println("Maximum weight that can be obtained is: " + maxWeight);
    }
    
    public static int findMaxWeight(List<Interval> intervals) {
        Collections.sort(intervals);
        int[] maxWeight = new int[intervals.size()];
        maxWeight[0] = intervals.get(0).weight;
        for (int i = 1; i < intervals.size(); i++) {
            int weight = intervals.get(i).weight;
            int compatible = findCompatible(intervals, i);
            if (compatible != -1) {
                weight += maxWeight[compatible];
            }
            maxWeight[i] = Math.max(weight, maxWeight[i-1]);
        }
        return maxWeight[intervals.size()-1];
    }
    
    public static int findCompatible(List<Interval> intervals, int index) {
        int low = 0;
        int high = index-1;
        while (low <= high) {
            int mid = (low+high)/2;
            if (intervals.get(mid).end <= intervals.get(index).start) {
                if (intervals.get(mid+1).end <= intervals.get(index).start) {
                    low = mid+1;
                } else {
                    return mid;
                }
            } else {
                high = mid-1;
            }
        }
        return -1;
    }
}